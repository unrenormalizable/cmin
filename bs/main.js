export const TOK_ERROR = 0
export const TOK_EOF = 1
export const TOK_WHITESPACE = 2
export const TOK_KEYWORD = 3
export const TOK_IDENTIFIER = 4
export const TOK_SYMBOL = 4

function getCurrChar(text, state) {
  return state.pos + 0 < text.length ? text[state.pos + 0] : null
}

function getNextChar(text, state) {
  return state.pos + 1 < text.length ? text[state.pos + 1] : null
}

function isWhitespace(char) {
  return char !== null && char.match(/\s/)
}

function isAlphabetic(char) {
  return char !== null && char.match(/[A-Za-z_]/)
}

function isNumeric(char) {
  return char !== null && char.match(/[0-9]/)
}

function isAlphanumeric(char) {
  return isAlphabetic(char) || isNumeric(char)
}

function isSymbol(char) {
  return char !== null && char.match(/[\(\)\{\}]/)
}

function isKeyword(str) {
  return (str === 'fn')
}

function scan(text, state) {
  state = { ...state }

  // WHITESPACE token.
  while (isWhitespace(getCurrChar(text, state))) {
    while (true) {
      const currChar = getCurrChar(text, state)
      const nextChar = getNextChar(text, state)

      if (currChar === '\t' || currChar === ' ') {
        state.pos++
        state.column++
        continue
      }

      if ((currChar === '\r' && nextChar === '\n')) {
        state.pos += 2
        state.line++
        state.column = 1
        continue
      }

      if (currChar === '\r' || currChar === '\n') {
        state.pos++
        state.line++
        state.column = 1
        continue
      }

      break
    }
  }

  // EOF token.
  if (getCurrChar(text, state) === null) {
    return { token: { id: TOK_EOF, value: null, location: state }, state }
  }

  // KEYWORD or IDENTIFIER token.
  if (isAlphabetic(getCurrChar(text, state))) {
    const location = { ...state }
    while (isAlphanumeric(getCurrChar(text, state))) {
      state.pos++
      state.column++
    }

    const value = text.slice(location.pos, state.pos)
    const id = isKeyword(value) ? TOK_KEYWORD : TOK_IDENTIFIER
    return { token: { id, value, location }, state }
  }

  // SYMBOL token.
  if (isSymbol(getCurrChar(text, state))) {
    const location = { ...state }
    state.pos++
    state.column++
    return { token: { id: TOK_SYMBOL, value: text.slice(location.pos, state.pos), location }, state }
  }

  // ERROR token.
  return { token: { id: TOK_ERROR, value: null, location: state }, state }
}

export function scanAll(text) {
  let ret = { state: { pos: 0, line: 1, column: 1 } }
  const tokens = []
  while (true) {
    ret = scan(text, ret.state)
    tokens.push(ret.token)
    if (ret.token.id === TOK_EOF || ret.token.id === TOK_ERROR) {
      break
    }
  }
  return tokens
}

export const AST_NODE_PROGRAM = 0
export const AST_NODE_FUNCTION = 0

function parseFunction(tokens, state) {
  state = { ...state }

  state.pos += 9
  return {
    ast: { id: AST_NODE_FUNCTION, name: tokens[1].value, return: 'void', location: tokens[1].location, statements: [] },
    state,
    errors: [],
  }
}

function parseProgram(tokens, state) {
  state = { ...state }
  const errors = []
  const ast = { id: AST_NODE_PROGRAM, functions: [] }

  if (tokens[state.pos].id === TOK_ERROR) {
    errors.push({ location: tokens[state.pos].location, message: 'scanner error token' })
    state.pos += 2
    return { ast, state, errors }
  }

  while (state.pos < tokens.length && tokens[state.pos].id === TOK_KEYWORD && tokens[state.pos].value === 'fn') {
    const fnRet = parseFunction(tokens, state)
    ast.functions.push(fnRet.ast)
    errors.push(...fnRet.errors)
    state = fnRet.state
  }

  state.pos++
  return { ast, state, errors }
}

export function parse(text) {
  const tokens = scanAll(text)
  if (tokens.length === 0) {
    return null
  }

  const state = { pos: 0 }
  return parseProgram(tokens, state)
}

export function analyse(program) {
  const warnings = []
  const errors = []

  let foundMain = false
  for (let i = 0; i < program.ast.functions.length; i++) {
    if (program.ast.functions[i].name === '__puts') {
      warnings.push({ location: program.ast.functions[i].location, message: 'missing main function' })
    }
    if (program.ast.functions[i].name === 'main') {
      foundMain = true
    }
  }
  if (!foundMain) {
    errors.push({ location: { line: 0, column: 0 }, message: 'missing main function' })
  }

  return { warnings, errors }
}

export function emit(program) {
  const warnings = []
  const errors = []

  let output = 'source_filename = "__todo__.cmin"\n\n'

  for (let i = 0; i < program.ast.functions.length; i++) {
    const fn = program.ast.functions[i]
    output += `define ${fn.return} @${fn.name}() {\n`
    output += `  ret ${fn.return}\n`
    output += '}\n'
  }

  return { output, warnings, errors }
}

// Learn more at https://docs.deno.com/runtime/manual/examples/module_metadata#concepts
if (import.meta.main) {
  const text = 'fn main() {\n}'
  const ret = parse(text)
  console.log(ret)
}
