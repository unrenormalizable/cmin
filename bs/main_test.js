import { assertSnapshot } from '@std/testing/snapshot'
import * as main from './main.js'

// Build executable using:
// & "C:\Program Files\LLVM\bin\clang.exe"  -Wno-override-module -o example.exe example.ll -fuse-ld=lld "-Wl,/SUBSYSTEM:CONSOLE,/DEFAULTLIB:libcmt,/DEBUG:FULL"

Deno.test({
  name: 'scan: empty file',
  fn: async (tc) => {
    const text = ''
    const tokens = main.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'scan: random error token',
  fn: async (tc) => {
    const text = '   \r#  \r\n  \n  '
    const tokens = main.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'scan: empty file with whitespace',
  fn: async (tc) => {
    const text = '   \r\t  \r\n  \n  '
    const tokens = main.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'scan: trivial main function',
  fn: async (tc) => {
    const text = 'fn main() {\n}'
    const tokens = main.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'parse: empty file',
  fn: async (tc) => {
    const text = ''
    const ret = main.parse(text)

    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'parse: empty file with whitespace',
  fn: async (tc) => {
    const text = '   \r\t  \r\n  \n  '
    const ret = main.parse(text)

    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'parse: random scanner error',
  fn: async (tc) => {
    const text = '   \r#  \r\n  \n  '
    const ret = main.parse(text)

    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'parse: trivial main function',
  fn: async (tc) => {
    const text = 'fn main() {\n}'
    const ret = main.parse(text)

    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'analyze: override puts function',
  fn: async (tc) => {
    const text = 'fn __puts() {\n}'
    const ret = main.analyse(main.parse(text))

    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'emit: trivial main function',
  fn: async (tc) => {
    const text = 'fn main() {\n}'
    const program = main.parse(text)
    const ret = main.analyse(program)
    if (ret.errors.length !== 0) {
      throw new Error('analyzer returned errors', ret)
    }
    const output = main.emit(program)

    await assertSnapshot(tc, output, {})
  },
})
