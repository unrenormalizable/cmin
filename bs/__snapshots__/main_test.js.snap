export const snapshot = {};

snapshot[`scan: empty file 1`] = `
[
  {
    id: 1,
    location: {
      column: 1,
      line: 1,
      pos: 0,
    },
    value: null,
  },
]
`;

snapshot[`scan: random error token 1`] = `
[
  {
    id: 0,
    location: {
      column: 1,
      line: 2,
      pos: 4,
    },
    value: null,
  },
]
`;

snapshot[`scan: empty file with whitespace 1`] = `
[
  {
    id: 1,
    location: {
      column: 3,
      line: 4,
      pos: 14,
    },
    value: null,
  },
]
`;

snapshot[`scan: trivial main function 1`] = `
[
  {
    id: 3,
    location: {
      column: 1,
      line: 1,
      pos: 0,
    },
    value: "fn",
  },
  {
    id: 4,
    location: {
      column: 4,
      line: 1,
      pos: 3,
    },
    value: "main",
  },
  {
    id: 4,
    location: {
      column: 8,
      line: 1,
      pos: 7,
    },
    value: "(",
  },
  {
    id: 4,
    location: {
      column: 9,
      line: 1,
      pos: 8,
    },
    value: ")",
  },
  {
    id: 4,
    location: {
      column: 11,
      line: 1,
      pos: 10,
    },
    value: "{",
  },
  {
    id: 4,
    location: {
      column: 1,
      line: 2,
      pos: 12,
    },
    value: "}",
  },
  {
    id: 1,
    location: {
      column: 2,
      line: 2,
      pos: 13,
    },
    value: null,
  },
]
`;

snapshot[`parse: empty file 1`] = `
{
  ast: {
    functions: [],
    id: 0,
  },
  errors: [],
  state: {
    pos: 1,
  },
}
`;

snapshot[`parse: empty file with whitespace 1`] = `
{
  ast: {
    functions: [],
    id: 0,
  },
  errors: [],
  state: {
    pos: 1,
  },
}
`;

snapshot[`parse: random scanner error 1`] = `
{
  ast: {
    functions: [],
    id: 0,
  },
  errors: [
    {
      location: {
        column: 1,
        line: 2,
        pos: 4,
      },
      message: "scanner error token",
    },
  ],
  state: {
    pos: 2,
  },
}
`;

snapshot[`parse: trivial main function 1`] = `
{
  ast: {
    functions: [
      {
        id: 0,
        location: {
          column: 4,
          line: 1,
          pos: 3,
        },
        name: "main",
        return: "void",
        statements: [],
      },
    ],
    id: 0,
  },
  errors: [],
  state: {
    pos: 10,
  },
}
`;

snapshot[`analyze: override puts function 1`] = `
{
  errors: [
    {
      location: {
        column: 0,
        line: 0,
      },
      message: "missing main function",
    },
  ],
  warnings: [
    {
      location: {
        column: 4,
        line: 1,
        pos: 3,
      },
      message: "missing main function",
    },
  ],
}
`;

snapshot[`emit: trivial main function 1`] = `
{
  errors: [],
  output: 'source_filename = "__todo__.cmin"

define void @main() {
  ret void
}
',
  warnings: [],
}
`;
